examples/load-reporting-service/server/lrs_server.go

```diff
@@ -0,0 +1,123 @@
+package server
+
+import (
+	"context"
+	core "github.com/envoyproxy/go-control-plane/envoy/api/v2/core"
+	gcpLoadStats "github.com/envoyproxy/go-control-plane/envoy/service/load_stats/v2"
+	"github.com/golang/protobuf/ptypes/duration"
+	"google.golang.org/grpc"
+	"log"
+)
+
+// Server handling Load Stats communication
+type Server interface {
+	gcpLoadStats.LoadReportingServiceServer
+	SendResponse(cluster string, upstreamCluster []string, frequency int64)
+}
+
+func NewServer(ctx context.Context) Server {
+	return &server{ctx: ctx, lrsCache: make(map[string]map[string]NodeMetadata)}
+}
+
+type server struct {
+	ctx context.Context
+	// This cache stores stream objects (and Node data) for every node (within a cluster) upon connection
+	lrsCache map[string]map[string]NodeMetadata
+}
+
+// Struct to hold stream object and node details
+type NodeMetadata struct {
+	stream stream
+	node   *core.Node
+}
+
+type stream interface {
+	grpc.ServerStream
+
+	Send(*gcpLoadStats.LoadStatsResponse) error
+	Recv() (*gcpLoadStats.LoadStatsRequest, error)
+}
+
+// Handles incoming stream connections and LoadStatsRequests
+func (s *server) StreamLoadStats(stream gcpLoadStats.LoadReportingService_StreamLoadStatsServer) error {
+	reqCh := make(chan *gcpLoadStats.LoadStatsRequest)
+
+	// goroutine to handle incoming LoadStatsRequests
+	go func() {
+		for {
+			req, err := stream.Recv()
+			if err != nil {
+				close(reqCh)
+				return
+			}
+			reqCh <- req
+		}
+	}()
+
+	for {
+		select {
+		case <-s.ctx.Done():
```

Comment:
when will this case happen? Errors from the stream will be delivered on .Recv(). AFAIC, nobody sets a timeout on the context or ever cancels it.